hasher-matcher-actioner/hmalib/common/content_models.py

```diff
@@ -0,0 +1,222 @@
+        return result
+
+    @classmethod
+    def get_from_content_id(
+        cls,
+        table: Table,
+        content_id: str,
+    ) -> t.List["ActionEventRecord"]:
+
+        items = table.query(
+            KeyConditionExpression=Key("PK").eq(
+                cls.get_dynamodb_content_key(content_id)
+            )
+            & Key("SK").begins_with(cls.ACTION_TIME_PREFIX),
+            ProjectionExpression=cls.DEFAULT_PROJ_EXP,
+        ).get("Items", [])
+
+        return cls._result_item_to_action_record(items)
+
+    @classmethod
+    def _result_item_to_action_record(
+        cls,
+        items: t.List[t.Dict],
+    ) -> t.List["ActionEventRecord"]:
+        return [
+            ActionEventRecord(
+                content_id=cls.remove_content_key_prefix(
+                    item["PK"],
+                ),
+                time_performed=datetime.datetime.fromisoformat(
+                    item["SK"][len(cls.ACTION_TIME_PREFIX) :]
+                ),
+                action_label=item["ActionLabel"],
+                action_rules=item["ActionRules"],
+            )
+            for item in items
+        ]
+
+
+@dataclass
+class ContentObject(ContentObjectBase, JSONifiable):
+    """
+    Content object that stores the values related to a specific
+    content_id and content_type.
+    Note/TODO: Something like an override flag should be added to the submit API
+    as ContentRef equality check on bytes is impractical so instead
+    something like:
+     "This content_id exists, [x] overwrite? (I know what you're doing)"
+    is likely better.
+    """
+
+    CONTENT_TYPE_PREFIX = "content_type#"
+
+    content_type: str
+    content_ref: str
+    content_ref_type: str
+    additional_fields: t.Optional[t.Set]
+    submissions: t.List
+    created_on: datetime.datetime
+    updated_at: datetime.datetime
+
+    @staticmethod
+    def get_dynamodb_content_type_key(content_type: str) -> str:
+        return f"{ContentObject.CONTENT_TYPE_PREFIX}{content_type}"
+
+    def to_json(self) -> t.Dict:
+        """
+        TODO docstring
+        """
+        result = asdict(self)
+        result.update(
+            additional_fields=list(
+                self.additional_fields if self.additional_fields else set()
+            ),
+            submissions=[s.isoformat() for s in self.submissions],
+            created_on=self.created_on.isoformat(),
+            updated_at=self.updated_at.isoformat(),
+        )
+        return result
+
+    def write_to_table(self, table: Table):
+        """
+        Write operations for this object need to be special cased (to avoid overwritting)
+        Therefore we do not implement `to_dynamodb_item`
+
+        If you're curious it would ~look like this:
+        def to_dynamodb_item(self) -> dict:
+            return {
+                "PK": self.get_dynamodb_content_key(self.content_id),
+                "SK": self.get_dynamodb_content_type_key(self.content_type),
```

Comment:
Can you explain why content_type may be different for the same content_key?

I know we have discussed this in the past, but might help others who read this.

Reply:
I mean, I don't think it should (same id w/different type) be honestly but we don't outright enforce it not being the case currently. The same is true for signal id and signal type. 

Do you mean to call out how a piece of content can produce multiple signals and they would share a content id but be different types?

Reply:
The SK being content type makes me feel that we are expecting multiple items in this partition.

However, I can now see that this is more due to Content Items cohabiting a table which has PK and SK.

So, tell me if this is accurate. For HMADataStore

| PartitionKey | SortKey | ItemType |
|-------------|-----------|-----------|
|`c#images/129901.jpg`| `type#pdq` | `PipelinePDQHashRecord`|
|`c#images/129901.jpg`| `s#te#<indicator_id>` | `PDQMatchRecord`|
|`c#images/129901.jpg`| `content_type#photo` | `ContentObject`|
|`c#images/129901.jpg`| `action_time#2021-05-01.....` | `ActionEventRecord`|

I think we are doing a good job of naming right now, but for PipelinePDQHashRecord and PDQMatchRecord, it is not immediately apparent (looking at PK and SK alone) that we are looking at items of that type.

For the 3rd and 4th row, it is abundantly clear that I will receive an object of type ContentObject and ActionEventRecord by looking at the PK and SK design. But for row 1 and 2, it is less clear.

We could have the SK for hash records be `hash_type#pdq`. Similarly, for match records, it could be `signal_match#te#<indicator>` or something similar.

Anyway, all said and done, I think we need to be documenting this somewhere.

Reply:
Agree and think it is accurate. The need to document is potentially issue worthy at this point.